### PR TITLE
Fix S3603 Bug and FN: NRE for local method with qualified attributes

### LINKED ISSUE #5117: Fix S3603 Bug and FN: NRE for local method with qualified attributes
### Description

When having a local function, with an attribute, PureAttributeOnVoidMethod  has a NullReferenceException.
`CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.PureAttributeOnVoidMethod' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.`

### Repro steps

``` csharp
namespace ClassLibrary1
{
    public static class Class1
    {
        public static void Function1()
        {
            DoSomething();

            [System.Diagnostics.CodeAnalysis.SuppressMessage("category", "checkId")]
            static void DoSomething()
            {
                // nothing to see here
            }
        }
    }
}
```

### Known workarounds

Use `pragma warning disable` if this is a `SuppressMessage`.

### Related information

* SonarAnalyzer.CSharp: 8.32.0.39516
* Visual Studio 2022
* dotnet sdk: 6.0.100
* Windows 11

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5117

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/PureAttributeOnVoidMethod.cs [hunk 1]
```diff
@@ -50,7 +51,8 @@ protected override void Initialize(SonarAnalysisContext context)
                 SyntaxKindEx.LocalFunctionStatement);
         }
 
-        private static bool IsPureAttribute(string attributeName) =>
-            attributeName.Equals("Pure") || attributeName.Equals("PureAttribute");
+        private static bool IsPureAttribute(AttributeSyntax attribute) =>
+            attribute.Name.GetIdentifier() is { } name
+            && (name.Identifier.ValueText == "Pure" || name.Identifier.ValueText == "PureAttribute");
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/PureAttributeOnVoidMethod.cs:56]: Don't we want to check here also the actual symbol? (Otherwise we will have false positives with user-defined attributes named `Pure`)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/ExpandingArchives.cs [hunk 2]
```diff
@@ -35,6 +35,7 @@ public void ExtractEntry(ZipArchiveEntry entry)
             Stream stream;
 
             entry.ExtractToFile(""); // Noncompliant
+            entry?.ExtractToFile(""); // Noncompliant
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/ExpandingArchives.cs:38]: Is this change intended?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/JwtSigned.cs [hunk 3]
```diff
@@ -22,6 +22,7 @@ void DecodingWithDecoder(JwtDecoder decoder)
             string decoded;
             decoded = decoder.Decode(invalidToken, secret, true);
             decoded = decoder.Decode(invalidToken, secret, false); // Noncompliant {{Use only strong cipher algorithms when verifying the signature of this JWT.}}
+            decoded = decoder?.Decode(invalidToken, secret, false); // Noncompliant
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/JwtSigned.cs:25]: Same comment also here.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/ExpandingArchives.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/ExpandingArchives.cs:38]: Yes, it's the [Bump Coverage](https://github.com/SonarSource/sonar-dotnet/pull/5165/commits/319b0e759ccedb5d3240b3366af5270ed3a242b4) commit. `MemberBindingExpressionSyntax` case was not covered in the handler.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/JwtSigned.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/JwtSigned.cs:25]: Same as above (just to have the same thing covered in two rules in case we delete one of them in the far future)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/PureAttributeOnVoidMethod.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/PureAttributeOnVoidMethod.cs:56]: `InvalidPureDataAttributeUsage` does that inside (see `GetPureAttribute` in the base class)

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/PureAttributeOnVoidMethod.cs:56]: I've added UT

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!
